compiler/src/dotty/tools/dotc/util/Signatures.scala

```diff
@@ -41,109 +43,238 @@ object Signatures {
+    case other => other.stripAnnots
 
-    val appliedDenot = fun.symbol.asSingleDenotation.mapInfo(_ => fun.tpe) :: Nil
-    (activeParameter, 0, appliedDenot)
+  /**
+   * Get index of currently edited parameter in unapply context.
+   *
+   * @param patterns Currently applied patterns for unapply method
+   * @param span The position of the cursor
+   * @param maximumParams Number of parameters taken by unapply method
+   * @return Index of currently edited parameter
+   */
+  private def unapplyParameterIndex(patterns: List[tpd.Tree], span: Span, maximumParams: Int)(using Context): Int =
+    val patternPosition = patterns.indexWhere(_.span.contains(span))
+    (patternPosition, patterns.length) match
+      case (-1, 0) => 0 // there are no patterns yet so it must be first one
+      case (-1, pos) => -1 // there are patterns, we must be outside range so we set no active parameter
+      case _ => (maximumParams - 1) min patternPosition max 0 // handle unapplySeq to always highlight Seq[A] on elements
 
   private def isTuple(tree: tpd.Tree)(using Context): Boolean =
-    ctx.definitions.isTupleClass(tree.symbol.owner.companionClass)
-
-  private def extractParamTypess(resultType: Type)(using Context): List[List[Type]] =
-    resultType match {
-      // Reference to a type which is not a type class
-      case ref: TypeRef if !ref.symbol.isPrimitiveValueClass =>
-        getExtractorMembers(ref)
-      // Option or Some applied type. There is special syntax for multiple returned arguments:
-      //   Option[TupleN] and Option[Seq],
-      // We are not intrested in them, instead we extract proper type parameters from the Option type parameter.
-      case AppliedType(TypeRef(_, cls), (appliedType @ AppliedType(tycon, args)) :: Nil)
-          if (cls == ctx.definitions.OptionClass || cls == ctx.definitions.SomeClass) =>
-        tycon match
-          case TypeRef(_, cls) if cls == ctx.definitions.SeqClass => List(List(appliedType))
-          case _ => List(args)
-      // Applied type extractor. We must extract from applied type to retain type parameters
-      case appliedType: AppliedType => getExtractorMembers(appliedType)
-      // This is necessary to extract proper result type as unapply can return other methods eg. apply
-      case MethodTpe(_, _, resultType) =>
-        extractParamTypess(resultType.widenDealias)
-      case _ =>
-        Nil
-    }
+    tree.symbol != NoSymbol && ctx.definitions.isTupleClass(tree.symbol.owner.companionClass)
+
+  /**
+   * Get unapply method result type omiting unknown types and another method calls.
+   *
+   * @param fun Unapply tree
+   * @return Proper unapply method type after extracting result from method types and omiting unknown types.
+   */
+  private def unapplyMethodResult(fun: tpd.Tree)(using Context): Type =
+    val typeWithoutBinds = fun match
+      case TypeApply(_, Bind(_, _) :: _) => fun.symbol.asSingleDenotation.info
```

Comment:
If there is no type applied, it is bound by Bind tree. We should skip it and return original function return type without it's type applied.
```scala
case class Two[T](a: T, b: T)
object Main {
  (null: Any) match {
    case Two(@@) =>
  }
}
```
Now returns `(a: T, b: T)` instead of `(a: T$1, b: T$1)`